docs/internal/contributing/README.md

```diff
@@ -105,6 +105,22 @@ You have to commit the changes to `go.mod` and `go.sum` before submitting the pu
 
 Please see the dedicated "[Design patterns and Code conventions](design-patterns-and-conventions.md)" page.
 
+## ⚠️ Unsafe memory tricks
+
+For performance, buffers used for decompressing gRPC requests into are kept in a pool shared between requests.
+
+Also, when unmarshaling those byte buffers into structured Go values, we construct strings that aren't actually references to immutable memory, but to those same shared buffers. The codebase refers to this trick as `yoloString`, and in some places (but not exhaustively), such strings are denoted as type `unsafeMutableString`, which is an alias of `string`.
```

Comment:
Can you point to an example where we construct a `yoloString` into a _decompression_ buffer? I just know that we have `yoloString`s into _deserialization_ buffers.

Reply:
Ultimately, those uncompressed buffers are passed to [LabelAdapter.Unmarshal](https://github.com/grafana/mimir/blob/8aeca96620cab68143176b1454deb4c6fdd934f1/pkg/mimirpb/timeseries.go#L362), where we call `yoloString` to get strings pointing into the source buffers.

Reply:
`yoloString` in `LabelAdapter.Unmarshal` references the `dAtA` argument to `LabelAdapter.Unmarshal`. Isn't `dAtA` a gRPC pooled buffer?

Reply:
`LabelAdapter` is wrapped by `PreallocWriteRequest`.

Because `PreallocWriteRequest` implements `proto.Unmarshaler`, `util.ProtoReader` wll (after having called `decompressRequest` into a buffer from `RequestBuffers`) call `PreallocWriteRequest.Unmarshal` directly, [here](https://github.com/grafana/mimir/blob/ec4af8e88b6b0898276f5a413ec8b037bf1f8421/pkg/util/http.go#L161-L165).

I think by doing this we're bypassing gRPC's internal pooling entirely.

Reply:
Ah that explains it, thanks @tcard. I don't even remember seeing that trick.